src/hotspot/share/oops/inlineKlass.cpp

```diff
@@ -153,6 +153,9 @@ bool InlineKlass::flat_array() {
   if (!UseFlatArray) {
     return false;
   }
+  if (VectorSupport::is_vector_payload_mf(this) || VectorSupport::is_vector(this)) {
+    return false;
+  }
```

Comment:
Why do you close the array flatten for vectors?

Reply:
Array of vector may be heterogeneous in nature where each element points to a vector of different shape, thus array flatten do not make much sense to vectors. In additional during type resolution flat_array flag is set for imbalanced Phi (having different type of inputs e.g. VectorPayload64 and VectorPayload128) even though Phi's type is lower common ancestor type of participating inputs i.e., VectorPayload which itself is not an inlinetype, this leads to assertion failures in downstream flow.

https://github.com/openjdk/valhalla/blob/lworld/src/hotspot/share/opto/type.cpp#L4636

Reply:
Thanks for the explanation! Seems it also causes the jvm crash in `VectorXXXConversionTests`, but I didn't figure out the root cause yet.  Is it possible supporting the flat array for vector objects with some fixes in future? We may lose the performance with this change, right?

Reply:
Since we never expose concrete vector types to users hence its only possible to create an array of abstract vector references which can be heterogeneous. This is different from an array of value types which are implicitly final classes and may be flattened into a compact layout since such an array will be homogeneous in nature.

Reply:
Yeah, I understand. Maybe we could find other more general ways to forbit such kind of array flatten like vector objects, instead of hard coding the `vector/vector_payload` classes. We can revisit this in future. WDYT?